juju/juju
# [JUJU-3332] Support JWT login tokens for controller and model access

Add support for JWT in the login request.
We support "controller" permissions (superuser, login) and "model" permissions (admin, write, read)

Use the controller config `login-token-refresh-url` to specify the endpoint to refresh the token public key from.
This can take 2 forms:
1.  `https://my.host`
2. `https://my.host/.wellknown/jwks.json`

In the first case, we append `/.wellknown/jwks.json`.

The token subject must be a user tag, like `user-fred`. ie who the request is being made on behalf of.

## QA steps

For now, just ensure that a bootstrap without `login-token-refresh-url` works as before.

The new functionality needs to be tested with JAAS, where the controller is started like so:

`juju bootstrap --config login-token-refresh-url=https://my.host ...`
